### PR TITLE
refactor document builder.

### PR DESCRIPTION
**Pull request by vwxyzh**
@chenkennt @hellosnow @ansyral @qinezh

### CODE REVIEW

**Comment by hellosnow** [comment]
:shipit:

**Comment by qinezh** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs [hunk 1]
```diff
@@ -427,45 +431,42 @@ private static void BuildArticle(IDocumentProcessor processor, HostService hostS
...
                     }
                 });
         }
 
-        private IEnumerable<ManifestItemWithContext> ExportManifest(InnerBuildContext buildContext, DocumentBuildContext context)
+        private IEnumerable<ManifestItemWithContext> ExportManifest(HostService hostService, DocumentBuildContext context)
         {
-            var hostService = buildContext.HostService;
-            var processor = buildContext.Processor;
-            var templateProcessor = buildContext.TemplateProcessor;
```
Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs]: Question, why move all these to HostService? what does a HostService instance mean?

Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs]: I don't want `build context` and `inner build context`, and I think no one know the different between them.
The `inner build context` is a named tuple for `host service`, `document processor` and `template processor`. And `host service` is an internal class, so merge them into that.
And now `host service` means the context for a document processor, and `IHostService` is a set of services providing for plug-ins.

Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs]: Thanks, i get it now. maybe later if we need to add more context related info, a HostServiceContext is still needed for better management

---
In reply to: [66018261](https://github.com/dotnet/docfx/pull/363#discussion_r66018261) [](ancestors = 66018261)

**Comment by ansyral** [comment]
:shipit:

**Pull request merged**